File: apps/api-backend/app/services/review_service.py
# Source: design.md TASK-B2-009 -- Review (spaced repetition) service
"""Review service: today's cards, rate card, summary."""

from typing import Any, Dict, List
from uuid import UUID

from sqlalchemy.ext.asyncio import AsyncSession

from app.repositories.review_card_repo import ReviewCardRepository


class ReviewService:
    def __init__(self, session: AsyncSession):
        self.session = session
        self.repo = ReviewCardRepository(session)

    async def get_today_cards(self, user_id: UUID) -> List[Dict[str, Any]]:
        """Get today's due review cards with vocabulary details. Source: T007."""
        # TODO: implement get_today_cards from T007
        pass

    async def rate_card(self, card_id: UUID, rating: int, user_id: UUID) -> Dict[str, Any]:
        """Submit rating for a review card.

        - Validate ownership
        - Validate rating 1-4 (REVIEW_002 on failure)
        - Call srs_service.schedule_card()
        - Persist updated card
        - Update VocabularyItem mastery_level if graduated
        Source: T007, TS004
        """
        # TODO: implement rate_card from T007
        pass

    async def get_summary(self, user_id: UUID) -> Dict[str, Any]:
        """Get review session summary: total due, reviewed, retention rate. Source: T007."""
        # TODO: implement get_summary from T007
        pass

File: apps/mobile-app/src/features/review/services/reviewService.ts
// Source: design.md Section 4.4 -- API calls for review endpoints
import { apiClient } from '../../../services/apiClient';
import { ReviewCardDTO, ReviewSummary } from '../../../types/api';

export const reviewService = {
  /**
   * GET /api/v1/reviews/today
   * Fetch today's due review cards.
   */
  async getTodayCards(): Promise<ReviewCardDTO[]> {
    // TODO: Implement API call
    const { data } = await apiClient.get('/reviews/today');
    return data.data;
  },

  /**
   * POST /api/v1/reviews/{card_id}/rate
   * Submit rating (1=again, 2=hard, 3=good, 4=easy).
   * FSRS algorithm recalculates next review date on backend.
   */
  async rateCard(cardId: string, rating: 1 | 2 | 3 | 4): Promise<void> {
    // TODO: Implement API call
    await apiClient.post(`/reviews/${cardId}/rate`, { rating });
  },

  /**
   * GET /api/v1/reviews/summary
   * Fetch review session summary.
   */
  async getSummary(): Promise<ReviewSummary> {
    // TODO: Implement API call
    const { data } = await apiClient.get('/reviews/summary');
    return data.data;
  },
};
